README.md
# Snowplow::EmrEtlRunner

## Introduction

Snowplow::EmrEtlRunner is a Ruby application (built with [Bundler][bundler]) to run Snowplow's Scalding-based Enrichment process on [Amazon Elastic MapReduce][amazon-emr] with minimum fuss.

## Find out more

| Technical Docs              | Setup Guide           | Roadmap & Contributing               |         
|-----------------------------|-----------------------|--------------------------------------|
| ![i1][techdocs-image]      | ![i2][setup-image]   | ![i3][roadmap-image]                |
| [Technical Docs][techdocs] | [Setup Guide][setup] | _coming soon_                        |

## Credits and thanks

Snowplow::EmrEtlRunner was primarily developed by [Alex Dean][alexanderdean] ([Snowplow Analytics][snowplow-analytics]), with very substantial contributions from [Michael Tibben][mtibben] ([99designs][99designs]). Huge thanks Michael!

EmrEtlRunner in turn depends heavily on [Rob Slifka][rslifka]'s excellent [Elasticity][elasticity] Ruby gem, which provides programmatic access to Amazon EMR. Big thanks to Rob for writing Elasticity!

## Copyright and license

Snowplow::EmrEtlRunner is copyright 2012-2020 Snowplow Analytics Ltd.

Licensed under the [Apache License, Version 2.0][license] (the "License");
you may not use this software except in compliance with the License.

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

[bundler]: http://gembundler.com/
[amazon-emr]: http://aws.amazon.com/elasticmapreduce/
[deploying-emr-etl-runner]: https://github.com/snowplow/snowplow/wiki/Deploying-EmrEtlRunner

[alexanderdean]: https://github.com/alexanderdean
[snowplow-analytics]: http://snowplowanalytics.com
[mtibben]: https://github.com/mtibben
[99designs]: http://99designs.com
[rslifka]: https://github.com/rslifka
[elasticity]: https://github.com/rslifka/elasticity

[license]: http://www.apache.org/licenses/LICENSE-2.0
[techdocs-image]: https://d3i6fms1cm1j0i.cloudfront.net/github/images/techdocs.png
[setup-image]: https://d3i6fms1cm1j0i.cloudfront.net/github/images/setup.png
[roadmap-image]: https://d3i6fms1cm1j0i.cloudfront.net/github/images/roadmap.png
[techdocs]: https://github.com/snowplow/snowplow/wiki/EmrEtlRunner
[setup]: https://github.com/snowplow/snowplow/wiki/setting-up-EmrEtlRunner


CHANGELOG
Version 1.0.4 (2020-08-21)
--------------------------
Not add/add S3DistCp for shredded JSONs and shredded TSVs depending on the configuration (#71)
Update config.yml with latest recommended settings (#79)

Version 1.0.3 (2020-07-30)
--------------------------
Disable retry on failure of the rdb_load step (#76)

Version 1.0.2 (2020-07-03)
--------------------------
Fix ParamContractError when copying shredded types to S3 (#74)

Version 1.0.1 (2020-07-02)
--------------------------
Fix undefined local variable bug (#72)

Version 1.0.0 (2020-06-10)
--------------------------
Handle step failures by resubmitting the jobflow (#63)

Version 0.38.1 (2020-06-08)
---------------------------
Fix default diplomat arguments (#65)

Version 0.38.0 (2020-06-05)
---------------------------
Catch and retry EMR connection issues (#4)
Bump upstream dependencies (#16)
Bump rest-client to 1.8.0 (#15)
Bump diplomat to 2.4.0 (#14)
Bump aws-sdk-s3 to 1.66.0 (#13)
Bump iglu-ruby-client to 0.2.0 (#12)
Bump snowplow-tracker to 0.6.1 (#11)
Bump awrence to 1.1.1 (#10)
Bump avro to 1.9.2 (#9)
Bump rspec to 3.9.0 (#8)
Bump jruby to 9.2.11.1 (#7)
Extend copyright to 2020 (#62)

Version 0.37.0 (2020-04-30)
---------------------------
Switch AMI bootstrap scripts to HTTPS (#4256)
Set ig.count (#4285)
Catch and retry EMR connection exceptions (#4290)

Version 0.36.0 (2019-12-03)
---------------------------
Add support to the spot market for core instances (#3487)

Version 0.35.0 (2019-09-12)
---------------------------
Add support for shredded TSV data (#4074)
Update spark_enrich version in config samples to 1.18.0 (#4091)

Version 0.34.3 (2019-07-17)
---------------------------
Update Contracts for get_failure_details (#4088)
Make sure all steps are successfully submitted in case of a transient cluster (#4092)

Version 0.34.2 (2019-05-17)
---------------------------
Extend backoff periods (#4049)
Limit requests to the EMR API (#4056)
Initialize S3::Aws constant (#4036)

Version 0.34.1 (2019-02-27)
---------------------------
Add exponential backoff when getting cluster statuses (#3995)
Update spark_enrich version in config sample to 1.17.0 (#3991)

Version 0.34.0 (2019-02-19)
---------------------------
Add support for running steps on a persistent EMR cluster (#3930)
Recover from request timeout (#3943)
Set timeout to 120 seconds (#3949)
Leverage compaction steps when copying to S3 (#3940)
Recover from S3 internal errors (#3950)
Rename EMR steps (#3925)
Scrub credentials from stderr (#2815)
Bump elasticity to 6.0.14 (#3948)
Extend copyright notice to 2019 (#3978)

Version 0.33.1 (2018-08-21)
---------------------------
Replace recursive functions by their iterative versions (#3866)
Retrieve correct latest run id when using s3a (#3871)
Update spark_enrich version in config samples to 1.16.0 (#3859)

Version 0.33.0 (2018-07-24)
---------------------------
Add ability to specify an EMR security configuration (#3798)
Handle SSE-S3 encrypted S3 buckets (#3456)
Replace Sluice by aws-sdk-s3 (#3524)
Add --ignore-lock-on-start option (#3537)
Check the processing folder for emptiness when resuming from enrich (#3803)
Make port in Snowplow monitoring configurable (#3236)
Make protocol in Snowplow monitoring configurable (#3791)
Update rdb_shredder version in config samples to 0.13.1 (#3790)
Update spark_enrich version in config samples to 1.14.0 (#3804)

Version 0.32.0 (2018-04-30)
---------------------------
Fail fast when trying to skip staging or enrich in stream enrich mode (#3726)
Factor out steps-generating function (#3718)
Uncompress enriched files when copying to HDFS (#3719)
Fix srcPattern for copying stream enriched data to HDFS (#3722)
Check if whole enriched.good is non-empty in stream enrich mode (#3717)
Update spark_enrich version in config.yml.sample to 1.13.0 (#3710)

Version 0.31.0 (2018-04-03)
---------------------------
Add ability to skip load_manifest_check (#3680)
Add CI/CD to update AMI bootstrap scripts (#3683)
Add stream_config.yml.sample (#3685)
Add support for shredding from Kinesis S3 Loader's enriched event output (#3606)
Add bootstrap action to prepare AMI 5.x for Snowplow (#3601)
Recover from RestClient::ServiceUnavailable when making status checks (#3539)
Recover from RestClient::RequestTimeout when making status checks (#3468)
Launch bootstrap action for AMI 5.x (#3609)
Pass processing manifest config to RDB Shredder (#3619)
Fail fast in build script (#3684)
Fail fast on duplicated storage target id (#3652)
Do not rescue on Exception (#3577)
Update spark_enrich version in config.yml.sample to 1.12.0 (#3566)

Version 0.30.0 (2017-12-18)
---------------------------
Add ability to skip RDB Loader consistency check (#3529)
Uncompress gzipped raw files when copying to HDFS (#3525)
Update spark_enrich version in config.yml.sample to 1.11.0 (#3002)

Version 0.29.0 (2017-11-13)
---------------------------
Reintroduce processing directory not empty no-op (#3458)
Retrieve the correct latest run ID during archive_shredded step (#3436)
Fix pagination issue when retrieving latest run id (#3434)
Update rdb_loader version in config.yml.sample to 0.14.0 (#3418)
Update rdb_shredder version in config.yml.sample to 0.13.0 (#3460)
Update spark_enrich version in config.yml.sample to 1.10.0 (#3461)
Bump AMI version in example config to 5.9.0 (#3465)
Force bundler 1.15.4 during CI/CD (#3493)

Version 0.28.0 (2017-09-11)
---------------------------
Release lock in case of no-op (#3396)
Treat archive_enriched and archive_shredded as separate steps (#3401)
Do not pass --skip shred to RDB Loader when skipping RDB Shredder (#3403)
If RDB Loader step hangs and is cancelled, logs are not retrieved (#3399)
Ensure appropriate log level for RDB logs (#3369)
Unlink downloaded RDB logs (#3363)
Do not try to download non-existent RDB loader log files (#3405)
Rescue the intermittent RestClient::SSLCertificateNotVerified error (#2572)
Pass GZIP compression argument to S3DistCp as "gz" not "gzip" (#3415)

lib/snowplow-emr-etl-runner/utils.rb
# Copyright (c) 2012-2020 Snowplow Analytics Ltd. All rights reserved.
#
# This program is licensed to you under the Apache License Version 2.0,
# and you may not use this file except in compliance with the Apache License Version 2.0.
# You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the Apache License Version 2.0 is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.

# Author::    Ben Fradet (mailto:support@snowplowanalytics.com)
# Copyright:: Copyright (c) 2012-2020 Snowplow Analytics Ltd
# License::   Apache License Version 2.0

require 'contracts'
require 'elasticity'

# Module with diverse utilities dealing with a few quirks in EmrEtlRunner
module Snowplow
  module EmrEtlRunner
    module Utils

      include Contracts

      # Builds the region-appropriate bucket name for Snowplow's hosted assets.
      # Has to be region-specific because of https://github.com/boto/botocore/issues/424
      #
      # Parameters:
      # +standard_bucket+:: Snowplow's hosted asset bucket
      # +bucket+:: the specified hosted assets bucket
      # +region+:: the AWS region to source hosted assets from
      Contract String, String, String => String
      def get_hosted_assets_bucket(standard_bucket, bucket, region)
        bucket = bucket.chomp('/')
        suffix = if !bucket.eql? standard_bucket or region.eql? "eu-west-1" then "" else "-#{region}" end
        "#{bucket}#{suffix}/"
      end

      # Get commons-codec version required by Spark Enrich for further replace
      # See: https://github.com/snowplow/snowplow/issues/2735
      Contract Maybe[String] => String
      def get_cc_version(she_version)
        if she_version.nil?
          "1.10"
        else
          she_version_normalized = Gem::Version.new(she_version)
          if she_version_normalized > Gem::Version.new("1.8.0")
            "1.10"
          else
            "1.5"
          end
        end
      end

      # Does this collector format represent CloudFront access logs?
      # Parameters:
      # +collector_format+:: collector format from config.yml, nil in case of stream mode
      Contract Maybe[String] => Bool
      def is_cloudfront_log(collector_format)
        if collector_format.nil?
          false
        else
          collector_format == "cloudfront" or
            collector_format.start_with?("tsv/com.amazon.aws.cloudfront/")
        end
      end

      # Does this collector format represent ndjson/urbanairship?
      # Parameters:
      # +collector_format+:: collector format from config.yml, nil in case of stream mode
      Contract Maybe[String] => Bool
      def is_ua_ndjson(collector_format)
        /^ndjson\/com\.urbanairship\.connect\/.+$/ === collector_format
      end

      # Returns the S3 endpoint for a given S3 region
      Contract String => String
      def get_s3_endpoint(s3_region)
        if s3_region == "us-east-1"
          "s3.amazonaws.com"
        else
          "s3-#{s3_region}.amazonaws.com"
        end
      end

      # Retrieves the paths of hadoop enrich, hadoop shred and hadoop elasticsearch
      Contract String, Maybe[String], String, String, String => AssetsHash
      def get_assets(assets_bucket, spark_enrich_version, rds_version, hadoop_elasticsearch_version, rdl_version)
        enrich_path_middle = if spark_enrich_version.nil?
          nil
        elsif is_spark_enrich(spark_enrich_version)
          'spark-enrich/snowplow-spark-enrich'
        else
          spark_enrich_version[0] == '0' ? 'hadoop-etl/snowplow-hadoop-etl' : 'scala-hadoop-enrich/snowplow-hadoop-enrich'
        end
        shred_path = if is_rdb_shredder(rds_version) then
          '4-storage/rdb-shredder/snowplow-rdb-shredder-'
        else
          '3-enrich/scala-hadoop-shred/snowplow-hadoop-shred-'
        end
        enrich_final = if enrich_path_middle.nil? then nil else "#{assets_bucket}3-enrich/#{enrich_path_middle}-#{spark_enrich_version}.jar" end
        {
          :enrich   => enrich_final,
          :shred    => "#{assets_bucket}#{shred_path}#{rds_version}.jar",
          :loader   => "#{assets_bucket}4-storage/rdb-loader/snowplow-rdb-loader-#{rdl_version}.jar",
          :elasticsearch => "#{assets_bucket}4-storage/hadoop-elasticsearch-sink/hadoop-elasticsearch-sink-#{hadoop_elasticsearch_version}.jar",
        }
      end

      # Check if the supplied shred version relates to the rdb-shredder or the
      # legacy scala-hadoop-shred.
      #
      # Parameters:
      # +shred_version+:: the specified shred version
      Contract String => Bool
      def is_rdb_shredder(shred_version)
        version = shred_version.split('.').map { |v| v.to_i }
        unless version.length == 3
          raise ArgumentError, 'The shred job version could not be parsed'
        end
        version[1] >= 12
      end

      # Check if the supplied enrich version relates to spark enrich or the legacy
      # scala-hadoop-enrich.
      #
      # Parameters:
      # +enrich_version+:: the specified enrich version
      Contract Maybe[String] => Bool
      def is_spark_enrich(enrich_version)
        if enrich_version.nil?
          false
        else
          version = enrich_version.split('.').map { |v| v.to_i }
          unless version.length == 3
            raise ArgumentError, 'The enrich job version could not be parsed'
          end
          version[0] >= 1 && version[1] >= 9
        end
      end

      # Returns a base64-encoded JSON containing an array of enrichment JSONs
      Contract ArrayOf[String] => String
      def build_enrichments_json(enrichments_array)
        enrichments_json_data = enrichments_array.map {|e| JSON.parse(e)}
        enrichments_json = {
          'schema' => 'iglu:com.snowplowanalytics.snowplow/enrichments/jsonschema/1-0-0',
          'data'   => enrichments_json_data
        }

        Base64.strict_encode64(enrichments_json.to_json)
      end

      Contract String => String
      def build_iglu_config_json(resolver)
        Base64.strict_encode64(resolver)
      end

      Contract Maybe[Iglu::SelfDescribingJson], Bool => Hash
      def build_duplicate_storage_json(target, snake_case=true)
        if target.nil?
          {}
        else
          encoded = Base64.strict_encode64(target.to_json.to_json)
          if snake_case
            { :duplicate_storage_config => encoded }
          else
            { 'duplicate-storage-config' => encoded }
          end
        end

      end

      # We need to partition our output buckets by run ID, buckets already have trailing slashes
      #
      # Parameters:
      # +folder+:: the folder to append a run ID folder to
      # +run_id+:: the run ID to append
      # +retain+:: set to false if this folder should be nillified
      Contract Maybe[String], String, Bool => Maybe[String]
      def partition_by_run(folder, run_id, retain=true)
        unless folder.nil?
          "#{folder}run=#{run_id}/" if retain
        else
          nil
        end
      end

      # Converts the output_compression configuration field to
      Contract Maybe[String] => ArrayOf[String]
      def output_codec_from_compression_format(compression_format)
        # those are the supported compression codecs
        if not compression_format.nil? and [ 'gzip', 'gz', 'lzo', 'snappy' ].include?(compression_format.downcase)
          downcased = compression_format.downcase
          format = [ 'gzip', 'gz' ].include?(downcased) ? 'gz' : downcased
          [ '--outputCodec', format ]
        else
          []
        end
      end

      # Adds a match all glob to the end of the path
      Contract String => String
      def glob_path(path)
        path = path.chomp('/')
        path.end_with?('/*') ? path : "#{path}/*"
      end

      DURATION_TOKENS = {
          "m" => (60),
          "h" => (60 * 60),
          "d" => (60 * 60 * 24),
          "w" => (60 * 60 * 24 * 7)
      }

      # Converts a duration string into a seconds integer
      # e.g. 5h 10m is converted into 18600
      Contract String => Integer
      def parse_duration(input)
        time = 0
        input.scan(/(\d+)(\w)/).each do |amount, measure|
          time += amount.to_i * DURATION_TOKENS[measure]
        end
        time
      end

      # Prettified string containing failure details
      # for this job flow.
      Contract String, Elasticity::ClusterStatus, ArrayOf[Elasticity::ClusterStepStatus] => String
      def get_failure_details(jobflow_id, cluster_status, cluster_step_status_for_run)
        [
          "EMR jobflow #{jobflow_id} failed, check Amazon EMR console and Hadoop logs for details (help: https://github.com/snowplow/snowplow/wiki/Troubleshooting-jobs-on-Elastic-MapReduce). Data files not archived.",
          "#{jobflow_id}: #{cluster_status.state} [#{cluster_status.last_state_change_reason}] ~ #{get_elapsed_time(cluster_status.ready_at, cluster_status.ended_at)} #{get_timespan(cluster_status.ready_at, cluster_status.ended_at)}"
        ].concat(cluster_step_status_for_run
            .sort { |a,b|
              self.class.nilable_spaceship(a.started_at, b.started_at)
            }
            .each_with_index
            .map { |s,i|
              " - #{i + 1}. #{s.name}: #{s.state} ~ #{get_elapsed_time(s.started_at, s.ended_at)} #{get_timespan(s.started_at, s.ended_at)}"
            })
          .join("\n")
      end

      # Gets the elapsed time in a
      # human-readable format.
      #
      # Parameters:
      # +start+:: start time
      # +_end+:: end time
      Contract Maybe[Time], Maybe[Time] => String
      def get_elapsed_time(start, _end)
        if start.nil? or _end.nil?
          "elapsed time n/a"
        else
          # Adapted from http://stackoverflow.com/a/19596579/255627
          seconds_diff = (start - _end).to_i.abs

          hours = seconds_diff / 3600
          seconds_diff -= hours * 3600

          minutes = seconds_diff / 60
          seconds_diff -= minutes * 60

          seconds = seconds_diff

          "#{hours.to_s.rjust(2, '0')}:#{minutes.to_s.rjust(2, '0')}:#{seconds.to_s.rjust(2, '0')}"
        end
      end

      # Gets the time span.
      #
      # Parameters:
      # +start+:: start time
      # +_end+:: end time
      Contract Maybe[Time], Maybe[Time] => String
      def get_timespan(start, _end)
        "[#{start} - #{_end}]"
      end

      def retry_connection_issues
        retries = 0
        handle_exception = ->(ex, description) {
          retries += 1
          delay = 4 ** retries + 30 # retry at 34, 46, 94, 286 seconds
          if retries < 5
            logger.warn "Got #{description} #{ex}, waiting #{delay} seconds before trying again"
            sleep(delay)
          else
            raise ex
          end
        }

        begin
          yield
        rescue SocketError => se
          handle_exception.call se, "socket error"
          retry
        rescue Errno::ECONNREFUSED => ref
          handle_exception.call ref, "connection refused"
          retry
        rescue Errno::ECONNRESET => res
          handle_exception.call res, "connection reset"
          retry
        rescue Errno::ETIMEDOUT => to
          handle_exception.call to, "connection timeout"
          retry
        rescue RestClient::InternalServerError => ise
          handle_exception.call ise, "internal server error"
          retry
        rescue Elasticity::ThrottlingException => te
          handle_exception.call te, "Elasticity throttling exception"
          retry
        rescue ArgumentError => ae
          handle_exception.call ae, "Elasticity argument error"
          retry
        rescue IOError => ioe
          handle_exception.call ioe, "IOError"
          retry
        rescue RestClient::SSLCertificateNotVerified => sce
          handle_exception.call sce, "RestClient::SSLCertificateNotVerified"
          retry
        rescue RestClient::RequestTimeout => rt
          handle_exception.call rt, "RestClient::RequestTimeout"
          retry
        rescue RestClient::ServiceUnavailable => su
          handle_exception.call su, "RestClient::ServiceUnavailable"
          retry
        rescue OpenSSL::SSL::SSLError => se
          handle_exception.call se, "OpenSSL::SSL::SSLError"
          retry
        end
      end

      # Tells if the job should copy shredded JSON files
      # False if shredder version >= shredder_tsv_version and tabularBlacklist is an empty array
      Contract Gem::Version, Gem::Version, ArrayOf[Iglu::SelfDescribingJson] => Bool
      def should_copy_shredded_JSONs(shredder_version, shredder_tsv_version, enriched_targets)
        tabularBlacklist = extract_tabular_blacklist(enriched_targets)
        shredder_version < shredder_tsv_version || (tabularBlacklist.nil? || !tabularBlacklist.empty?)
      end

      # Tells if the job should copy shredded TSV files
      # True if shredder version >= shredder_tsv_version and tabularBlacklist exists and is an array
      Contract Gem::Version, Gem::Version, ArrayOf[Iglu::SelfDescribingJson] => Bool
      def should_copy_shredded_TSVs(shredder_version, shredder_tsv_version, enriched_targets)
        tabularBlacklist = extract_tabular_blacklist(enriched_targets)
        shredder_version >= shredder_tsv_version && tabularBlacklist.kind_of?(Array)
      end

      # Extracts blackList tabular from redshift_config configuration file
      Contract ArrayOf[Iglu::SelfDescribingJson] => Maybe[ArrayOf[String]]
      def extract_tabular_blacklist(enriched_targets)
        redshift_target = enriched_targets.detect { |target|
          target.schema.name == 'redshift_config' && target.schema.version.model >= 4
        }
        if !redshift_target.nil? && !redshift_target.data[:blacklistTabular].nil? && redshift_target.data[:blacklistTabular].kind_of?(Array)
          redshift_target.data[:blacklistTabular]
        else
          nil
        end
      end
    end
  end
end
